adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -123,6 +126,139 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
     replaceRdd(tFn(rdd))
   }
 
+  def sorted: Boolean
+  // The partition map is structured as follows:
+  // The outer option is for whether or not there is a partition map.
+  //   - This is None in the case that we don't know the bounds on each 
+  //     partition.
+  // The Array is the length of the number of partitions.
+  // The inner option is in case there is no data on a partition.
+  // The (ReferenceRegion, ReferenceRegion) tuple contains the bounds of the 
+  //   partition, such that the lowest start is first and the highest end is
+  //   second.
+  def partitionMap: Option[Array[Option[(ReferenceRegion, ReferenceRegion)]]]
+
+  /**
+   * Repartitions all data in rdd and distributes it as evenly as possible
+   * into the number of partitions provided.
+   *
+   * @param partitions the number of partitions to repartition this.rdd into
+   * @return a new SortedGenomicRDDMixIn with the RDD partitioned evenly
+   */
+  private def evenlyRepartition(partitions: Int)(implicit tTag: ClassTag[T]): GenomicRDD[T, U] = {
+    require(sorted, "Cannot evenly repartition an unsorted RDD.")
+    val count = rdd.count
+    // we don't want a bunch of empty partitions, so we will just use count in
+    // the case the user wants more partitions than rdd records.
+    val finalPartitionNumber = Math.min(count, partitions)
+    // the average number of records on each node will help us evenly repartition
+    val average: Double = count.toDouble / finalPartitionNumber
+
+    // we already have a sorted rdd, so let's just move the data
+    // but we want to move it in blocks to maintain order
+    // zipWithIndex seems to be the cheapest way to guarantee this
+    val finalPartitionedRDD = rdd
+      .mapPartitionsWithIndex((idx, iter) => {
+        // here we get the destination partition number
+        // this will calculate so that each partition has a nearly
+        // equal amount of data on it
+        extractReferenceRegionsForPartition(iter, partitionMap.get(idx))
+      }, preservesPartitioning = true).zipWithIndex
+      .mapPartitionsWithIndex((idx, iter) => {
+        // divide the global index by the average to get the destination
+        // partition number
+        iter.map(_.swap).map(f => ((f._2._1, (f._1 / average).toInt), f._2._2))
+      }, preservesPartitioning = true)
+      .repartitionAndSortWithinPartitions(new ReferenceRegionRangePartitioner(finalPartitionNumber.toInt))
+
+    finalPartitionedRDD.cache()
+
+    val tempPartitionMap = finalPartitionedRDD.mapPartitions(iter =>
+      getRegionBoundsFromPartition(iter.map(f => (f._1._1, f._2))), preservesPartitioning = true).collect
+
+    // we want to try to avoid any None values in our final partition map
```

Comment:
`None`s in the `partitionMap` occur when there is no data on a partition. Currently, we don't enforce a rule that the number of partitions must be less than the amount of data. This was an explicit design decision because of the cost for the `count` operation for such an extreme edge case.